datafusion/datasource-parquet/src/metadata.rs

```diff
@@ -0,0 +1,564 @@
+        self
+    }
+
+    /// set file metadata cache
+    pub fn with_file_metadata_cache(
+        mut self,
+        file_metadata_cache: Option<Arc<dyn FileMetadataCache>>,
+    ) -> Self {
+        self.file_metadata_cache = file_metadata_cache;
+        self
+    }
+
+    /// Set timeunit to coerce INT96 timestamps to
+    pub fn with_coerce_int96(mut self, time_unit: Option<TimeUnit>) -> Self {
+        self.coerce_int96 = time_unit;
+        self
+    }
+
+    /// fetch parquet metadata
+    pub async fn fetch_metadata(&self) -> Result<Arc<ParquetMetaData>> {
+        let Self {
+            store,
+            object_meta,
+            metadata_size_hint,
+            decryption_properties,
+            file_metadata_cache,
+            coerce_int96: _,
+        } = self;
+
+        let fetch = ObjectStoreFetch::new(*store, object_meta);
+
+        // implementation to fetch parquet metadata
+        let cache_metadata =
+            !cfg!(feature = "parquet_encryption") || decryption_properties.is_none();
+
+        if cache_metadata {
+            if let Some(parquet_metadata) = file_metadata_cache
+                .as_ref()
+                .and_then(|file_metadata_cache| file_metadata_cache.get(object_meta))
+                .and_then(|file_metadata| {
+                    file_metadata
+                        .as_any()
+                        .downcast_ref::<CachedParquetMetaData>()
+                        .map(|cached_parquet_metadata| {
+                            Arc::clone(cached_parquet_metadata.parquet_metadata())
+                        })
+                })
+            {
+                return Ok(parquet_metadata);
+            }
+        }
+
+        let mut reader =
+            ParquetMetaDataReader::new().with_prefetch_hint(*metadata_size_hint);
+
+        #[cfg(feature = "parquet_encryption")]
+        if let Some(decryption_properties) = decryption_properties {
+            reader = reader.with_decryption_properties(Some(decryption_properties));
+        }
+
+        if cache_metadata && file_metadata_cache.is_some() {
+            // Need to retrieve the entire metadata for the caching to be effective.
+            reader = reader.with_page_indexes(true);
+        }
+
+        let metadata = Arc::new(
+            reader
+                .load_and_finish(fetch, object_meta.size)
+                .await
+                .map_err(DataFusionError::from)?,
+        );
+
+        if cache_metadata {
+            if let Some(file_metadata_cache) = file_metadata_cache {
+                file_metadata_cache.put(
+                    object_meta,
+                    Arc::new(CachedParquetMetaData::new(Arc::clone(&metadata))),
+                );
+            }
+        }
+
+        Ok(metadata)
+    }
+
+    /// Read and parse the schema of the Parquet file
+    pub async fn fetch_schema(&self) -> Result<Schema> {
+        let metadata = self.fetch_metadata().await?;
```

Comment:
Related to question above: how much work is fetching the schema? Does it also fetch row group stats? Page indexes?